src/main/java/graphql/execution/instrumentation/dataloader/DataLoaderDispatcherInstrumentationState.java

```diff
@@ -20,8 +31,11 @@ public DataLoaderDispatcherInstrumentationState(Logger log, DataLoaderRegistry d
         this.dataLoaderRegistry = dataLoaderRegistry;
         this.approach = new FieldLevelTrackingApproach(log, dataLoaderRegistry);
         this.state = approach.createState();
-        hasNoDataLoaders = dataLoaderRegistry.getKeys().isEmpty();
-
+        //
```

Comment:
Are we actually get that much performance improvements by caching this `hasNoDataLoaders` field in the constructor compared to calling `dataLoaderRegistry.getKeys().isEmpty()` in the method `boolean hasNoDataLoaders()` considering `HashSet` for example just does:  `size == 0`.  Looking into the code the only significant negative would be the unnecessary cloning of the keys `HashSet` each time which we could improve by adding a new method like `DataLoaderRegistry#isEmpty` or `DataLoaderRegistry#hasLoaders`.  This would remove the need for cloning the keys `HashSet` when calling `dataLoaderRegistry#keys` and then we are literally just replacing:

```
if (state.hasNoDataLoaders)
```

with an equivalent

```
if (dataLoaderRegistry.dataLoaders.size == 0)
```

The advantage over this approach is that the consumers of the library do not need to care about this `EMPTY_DATALOADER_REGISTRY`, they can just add the `DataLoaderRegistry` that they like and it should work as expected for this use case or they can register no registry and it would still work as they intended.

The only negative to this approach I can see is whether we want to expose an `isEmpty` method onto the `DataLoaderRegistry` class? Maybe there is a reason for this that I am missing?

Reply:
Another alternative design decision could be to convert the `DataLoaderRegistry` to an interface or make an interface that it implements. When you provide none you include a `EmptyDataLoaderRegistry` with all your performance improvements needs for the most common case.

If the developer just wants a static list of data loaders they could use an `ImmutableDataLoaderRegistry` which is immutable and cannot change and will still get the performance needs.

If you need something like this use case where we want to be able to change the data loaders during execution the developer could construct it with a `MutableDataLoaderRegistry` so it can be dynamically changed during the runtime. This would not be as performant with these checks but would allow for this edge case.